Communications/PacketBuffer.cs

```diff
@@ -0,0 +1,538 @@
+    /// </summary>
+    public class QueueBuffer : PacketBuffer
+    {
+        private readonly Queue<Packet> Queue; // Packet queue
+
+        public QueueBuffer()
+        {
+            Queue = new Queue<Packet>();
+        }
+
+        /// <summary>
+        /// Add a packet to buffer.
+        /// </summary>
+        /// 
+        /// <param name="Packet"> The packet to add. </param>
+        /// <param name="Priority"> Should always be zero because there's just one queue. </param>
+        /// 
+        /// <exception cref="ArgumentException"> If priority is not zero since there's just one queue. </exception>
+        public override void Enqueue(Packet Packet, int Priority = 0)
+        {
+            if (Priority != 0)
+                throw new ArgumentException("Leaky bucket controller doesn't have priority.");
+
+            lock (Queue)
+            {
+                Queue.Enqueue(Packet);
+            }
+        }
+
+        /// <summary>
+        /// Get next packet without removing it.
+        /// </summary>
+        /// 
+        /// <remarks>
+        /// It will return null if the buffer is empty instead of throwing an exception.
+        /// </remarks>
+        /// 
+        /// <returns> Next packet if buffer is not empty, or null otherwise. </returns>
+        public override Packet Peek()
+        {
+            lock (Queue)
+            {
+                try
+                {
+                    return Queue.Peek();
+                }
+                catch (InvalidOperationException)
+                {
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get next packet and remove it.
+        /// </summary>
+        /// 
+        /// <remarks>
+        /// It will return null if the buffer is empty instead of throwing an exception.
+        /// </remarks>
+        /// 
+        /// <returns> Next packet if buffer is not empty, or null otherwise. </returns>
+        public override Packet Dequeue()
+        {
+            lock (Queue)
+            {
+                try
+                {
+                    return Queue.Dequeue();
+                }
+                catch (InvalidOperationException)
+                {
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the total number of packets in the buffer.
+        /// </summary>
+        public override int Count
+        {
+            get
+            {
+                lock (Queue)
+                {
+                    return Queue.Count;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// This defines a priority buffer that dequeue the highest priority packet first.
+    /// </summary>
+    public class PriorityBuffer : PacketBuffer
+    {
+        private readonly PacketBuffer[] Buffers; // List of buffers of each priority
+        public readonly int NBuffers; // Number of buffers
```

Comment:
Wouldn't it make more sense to make this a get-only Property, so that there's no chance of it being incorrect?
Like this:
```
public int BufferQty
{
get { return Buffers.Length; }
}
```

Reply:
This is just a simple data structure. All fields are read-only, so it's more efficient to fetch Buffer.Length just once.